src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/SourceGenerators/SourceGeneratorProjectItem.cs

```diff
@@ -47,7 +48,19 @@ public override Stream Read()
             => throw new NotSupportedException("This API should not be invoked. We should instead be relying on " +
                 "the RazorSourceDocument associated with this item instead.");
 
-        public bool Equals(SourceGeneratorProjectItem? other) => other is not null && AdditionalText == other.AdditionalText;
+        public bool Equals(SourceGeneratorProjectItem? other)
+        {
+            if (ReferenceEquals(AdditionalText, other?.AdditionalText))
+            {
+                return true;
+            }
+
+            // in the compiler server when the generator driver cache is enabled the
+            // additional files are always different instances even if their content is the same
+            var thisHash = AdditionalText.GetText()?.GetChecksum() ?? [];
+            var otherHash = other?.AdditionalText.GetText()?.GetChecksum() ?? [];
+            return Enumerable.SequenceEqual(thisHash, otherHash);
```

Comment:
Should we then compare the actual text, or are we just going to accept our prize if we actually discover a collision?

Reply:
Heh, good question. I'm leaning towards just accept the prize, as I think the IDE also uses this to synchronize documents etc across OOP, so would suffer the same issue.